src/mjolnir/ferry_connections.cc

```diff
@@ -54,154 +55,185 @@ uint32_t ShortestPath(const uint32_t start_node_idx,
+
+        // Check if already in adj set - skip if cost is higher than prior path
+        if (node_status[endnode].set == kTemporary) {
+          uint32_t endnode_idx = node_status[endnode].index;
+          if (node_labels[endnode_idx].cost < cost) {
+            continue;
+          }
+        }
+
+        // Add to the node labels and adjacency set. Skip if this is a loop.
+        node_labels.emplace_back(cost, endnode, expand_node_idx);
+        node_status[endnode] = {kTemporary, nodelabel_index};
+        adjset.push({cost, nodelabel_index});
+        nodelabel_index++;
+      }
     }
-  }
 
-  // If only one label we have immediately found an edge with proper
-  // classification - or we cannot expand due to driveability
-  if (node_labels.size() == 1) {
-    LOG_DEBUG("Only 1 edge reclassified");
-    return 0;
-  }
+    // If only one label we have immediately found an edge with proper
+    // classification - or we cannot expand due to driveability
+    if (node_labels.size() == 1) {
+      LOG_DEBUG("Only 1 edge reclassified");
+      return 0;
+    }
 
-  // Trace shortest path backwards and upgrade edge classifications
-  uint32_t count = 0;
-  while (true) {
-    // Get the edge between this node and the predecessor
-    uint32_t idx = node_labels[index].node_index;
-    uint32_t pred_node = node_labels[index].pred_node_index;
-    auto expand_node_itr = nodes[idx];
-    auto bundle2 = collect_node_edges(expand_node_itr, nodes, edges);
-    for (auto& edge : bundle2.node_edges) {
-      if (edge.first.sourcenode_ == pred_node || edge.first.targetnode_ == pred_node) {
-        sequence<Edge>::iterator element = edges[edge.second];
-        auto update_edge = *element;
-        if (update_edge.attributes.importance > rc) {
-          update_edge.attributes.importance = rc;
-          update_edge.attributes.reclass_ferry = true;
-          element = update_edge;
-          count++;
+    // Trace shortest path backwards and upgrade edge classifications
+    // did we find all modes in the path?
+    uint32_t path_access = baldr::kVehicularAccess;
+    while (true) {
+      // Get the edge between this node and the predecessor
+      uint32_t idx = node_labels[label_idx].node_index;
+      uint32_t pred_node = node_labels[label_idx].pred_node_index;
+      auto expand_node_itr = nodes[idx];
+      auto bundle2 = collect_node_edges(expand_node_itr, nodes, edges);
+      for (auto& edge : bundle2.node_edges) {
+        bool forward = edge.first.sourcenode_ == pred_node;
+        if (forward || edge.first.targetnode_ == pred_node) {
+          sequence<Edge>::iterator element = edges[edge.second];
+          auto update_edge = *element;
+          if ((forward && inbound) || (!forward && !inbound)) {
+            path_access &= update_edge.rev_access;
+          } else if ((forward && !inbound) || (!forward && inbound)) {
+            path_access &= update_edge.fwd_access;
+          }
```

Comment:
remember the cumulative accesses for all vehicular modes along the found path with `path_access`